contrib/mesos/pkg/minion/server.go

```diff
@@ -243,6 +252,22 @@ func (ms *MinionServer) Run(hks hyperkube.Interface, _ []string) error {
 		log.Fatalf("No API client: %v", err)
 	}
 	ms.clientConfig = clientConfig
+	c := client.NewOrDie(clientConfig)
+
+	// create or update node object on apiserver. This must be done synchronously
+	// because the kubelet will refuse launching pods if the labels do not match.
+	labels := map[string]string{}
+	if ms.nodeLabelString != "" {
+		err = json.Unmarshal([]byte(ms.nodeLabelString), &labels)
+		if err != nil {
+			log.Fatalf("Invalid node-labels string: %v", err)
+		}
+	}
+	log.Infof("Creating or updating node %s with labels %v", ms.nodeHostName, labels)
+	_, err = node.CreateOrUpdate(c, ms.nodeHostName, labels)
```

Comment:
Actually, this PR adds complexity/dependencies to the minion controller:
- it's concerned with node host name (and needs a new flag)
- it's concerned with node labels (and needs a new flag)
- it's concerned with apiserver client config
- it becomes dependent upon apiserver

We should really avoid this. The current responsibilities of the minion
controller relate to cgroups, logging, and process management.

Instead of what I proposed (blocking in Run()) we could also just defer the
call to app.RunKubelet until initial registration completes. There are
other ways to skin this order-of-operations problem. I appreciate the need
to detangle/decouple the executor from the kubelet. Let's get creative :)


On Wed, Sep 16, 2015 at 8:40 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/minion/server.go
> <https://github.com/kubernetes/kubernetes/pull/13857#discussion_r39623888>
> :
>
> > @@ -243,6 +252,22 @@ func (ms *MinionServer) Run(hks hyperkube.Interface, _ []string) error {
> >  		log.Fatalf("No API client: %v", err)
> >  	}
> >  	ms.clientConfig = clientConfig
> > +	c := client.NewOrDie(clientConfig)
> > +
> > +	// create or update node object on apiserver. This must be done synchronously
> > +	// because the kubelet will refuse launching pods if the labels do not match.
> > +	labels := map[string]string{}
> > +	if ms.nodeLabelString != "" {
> > +		err = json.Unmarshal([]byte(ms.nodeLabelString), &labels)
> > +		if err != nil {
> > +			log.Fatalf("Invalid node-labels string: %v", err)
> > +		}
> > +	}
> > +	log.Infof("Creating or updating node %s with labels %v", ms.nodeHostName, labels)
> > +	_, err = node.CreateOrUpdate(c, ms.nodeHostName, labels)
>
> I don't like any explicit synchronization with the kubelet or executor.
> That's why this node creation takes place here. Reduces the complexity and
> dependencies.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13857/files#r39623888>.
>